pkg/controller/infrastructure/flow_reconciler.go

```diff
@@ -0,0 +1,396 @@
+	}
+	err = fctx.Delete(ctx)
+	if err != nil {
+		return err
+	}
+
+	tf, err := newTerraformer(f.log, f.restConfig, aws.TerraformerPurposeInfra, infra, f.disableProjectedTokenMount)
+	if err != nil {
+		return err
+	}
+	// TODO optimisation: check if cleanup is necessary
+	return CleanupTerraformerResources(ctx, tf)
+}
+
+// Restore implements the restoration of an infrastructure resource during the control plane migration.
+func (f *FlowReconciler) Restore(ctx context.Context, infra *extensionsv1alpha1.Infrastructure, cluster *controller.Cluster) error {
+	return f.Reconcile(ctx, infra, cluster)
+}
+
+func (f *FlowReconciler) infrastructureStateFromRaw(infra *extensionsv1alpha1.Infrastructure) (*awsapi.InfrastructureState, error) {
+	state := &awsapi.InfrastructureState{}
+	raw := infra.Status.State
+
+	if raw != nil {
+		jsonBytes, err := raw.MarshalJSON()
+		if err != nil {
+			return nil, err
+		}
+
+		// todo(ka): for now we won't use the actuator decoder because the flow state kind was registered as "FlowState" and not "InfrastructureState". So we
+		// shall use the simple json unmarshal for this release.
+		if err := json.Unmarshal(jsonBytes, state); err != nil {
+			return nil, err
+		}
+	}
+
+	return state, nil
+}
+
+func (f *FlowReconciler) migrateFromTerraform(ctx context.Context, infra *extensionsv1alpha1.Infrastructure) (*awsapi.InfrastructureState, error) {
+	var (
+		state = &awsapi.InfrastructureState{
+			Data: map[string]string{},
+		}
+	)
+	f.log.Info("starting terraform state migration")
+
+	// we want to prevent the deletion of Infrastructure CR if there may be still resources in the cloudprovider. We will initialize the data
+	// with a specific "marker" so that deletion attempts will not skip the deletion if we are certain that terraform had created infra resources
+	// in past reconciliation.
+	tf, err := newTerraformer(f.log, f.restConfig, aws.TerraformerPurposeInfra, infra, f.disableProjectedTokenMount)
+	if err != nil {
+		return nil, err
+	}
+
+	// nothing to do if state is empty
+	if tf.IsStateEmpty(ctx) {
+		return state, nil
+	}
+
+	// migrate state
+	infrastructureConfig, err := helper.InfrastructureConfigFromInfrastructure(infra)
+	if err != nil {
+		return nil, err
+	}
+	state, err = migrateTerraformStateToFlowState(infra.Status.State, infrastructureConfig.Networks.Zones)
+	if err != nil {
+		return nil, fmt.Errorf("migration from terraform state failed: %w", err)
+	}
+
+	// TODO duplication of computeProviderStatusFromFlowState and fctx.computeInfrastructureStatus
+	infrastructureStatus, err := computeProviderStatusFromFlowState(infrastructureConfig, state)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := infraflow.PatchProviderStatusAndState(ctx, f.client, infra, infrastructureStatus, &runtime.RawExtension{Object: state}, nil); err != nil {
+		return nil, fmt.Errorf("updating status state failed: %w", err)
+	}
+
+	f.log.Info("terraform state migrated successfully")
```

Comment:
The initial logging on migration is arguably useful. This one does look more like a leftover to me though.